proposals/0000-overloaded-bracket.rst

```diff
@@ -0,0 +1,394 @@
+   be generalised in the same manner.
+
+6. Type quotations are similarly generalised::
+
+   i :: Quote m => m (TExp (Int -> Int))
+   i = [|| \x -> x + 1 ||]
+
+
+Effect and Interactions
+-----------------------
+
+When making an interface more general it is important to think about whether it
+will affect type inference. If there are functions where we have to generalise
+the argument type but not the result then generalisation can result in
+ambiguity in the composition.
+
+It doesn't seem to me that there will be any problems with ambiguity here as
+the types of splices is not overloaded in the same manner.
+
+
+Interaction with Lift
+.....................
+
+The main breakage from this patch comes from modifying the type signature for
+``lift``.
+
+Instances defined using ``DeriveLift`` will continue to work because they are
+defined in terms of quotation brackets.
+
+Instances written in terms of the combinators from ``Language.Haskell.TH.Lib``
+will continue to work because these combinators will be generalised.
+
+Instances written in terms of ``Q`` will no longer work. For users to migrate
+an additional class ``LiftQ`` could be defined which has the old interface.
+This would mean users need to explicitly lift but there are likely only a few
+instances which fall into this category if any at all.
+
+Definition of Quote
+...................
+
+Richard observes that ``Language.Haskell.TH.Lib.Internal.numTyLit`` calls
+``fail`` from the ``Q`` monad. This call to ``fail`` can be replaced with
+a call to ``error``. It will still be executed at compile-time but with a
+potentially slightly worse error message. The alternative is to
+also add this effect to the ``Quote`` type class.
+
+Other Effects
+.............
+
+Vlad points out that you don't need to very strict about the types of
+expressions in splices. Each nested splice could have different constraints::
+
+      f :: Quasi m => m Exp
+      g :: MonadIO m => m Exp
+      [| putStrLn $(f) >> putStrLn $(g) |] :: (Quote m, Quasi m, MonadIO m) => m Exp
+
+If one of the nested splices has a specific type, for instance ``Q Exp``, then
+the type of the whole expression is fixed to be ``Q Exp``.
+
+
+Costs and Drawbacks
+-------------------
+
+* The generalisation of untyped brackets does not seem like it will cause
+  any significant breakage but it's hard to predict.
+* The modification to the ``Lift`` interface could cause user-written instances
+  to break but users should not define their own instances anyway.
+  ``DeriveLift`` is the blessed manner in which to define a ``Lift`` instance.
+
+Alternatives
+------------
+
+* The main alternative to the design would be to only require a ``Quote``
+  constraint when the quotation requires the ``newName`` effect. For example,
+  ``[| 5 |] :: Applicative m => m Exp``. I am opposed to this direction as it
+  breaks abstraction. The implementation detail of how ``[| 5 |]`` is desugared
+  leaks to the user.
```

Comment:
I don’t quite understand why this is worse: Why not have a less restrictive type if possible?

Can we draw an analogy, where `do e` does _not_ have a `Monad` constraint; the constrait only gets added if we add more lines. And `MonadFail` is only added if there is an incomplete pattern match? Seems analogous to this case.

Reply:
I guess this ties to Simon’s question

> What about quotes that do not need a name supply, like [| map inc [1,2,3] |]? The combinators for that don't even need a name supply, so do we get away with Applicative for them?

which is partly answers by this section, but maybe not sufficiently motivated?

Reply:
There's value in presenting a consistent API. Tagging all TH syntax combinators with the `Quote` constraint indicates that they are meant to be used together, much like the current definition of everything in terms of `Q` does.

There's also value in requesting the minimal required capabilities as it makes the API more flexible, but this usually becomes important when the capabilities are heavyweight somehow. A database connection might be heavyweight, but a name supply seems pretty light. So I'd probably lean towards using `Quote` consistently.

Reply:
@gridaphobe 
TH already has a consistent API, namely `Q`. The whole point of this proposal (as I see it) is to reduce the heavyweight `Q` to the lighter weight `Quote`. This will make the API inconstent, as reification and other such things will still need `Q`. Since this proposal is requesting that we reduce `Q` to its minimal required capabilities (at least for splices), why not go all the way and reduce it to `Applicative` in the cases where that is possible?